dev/annotate_source.py

```diff
@@ -0,0 +1,235 @@
+#!/usr/bin/python
+#
+# Copyright 2017 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import re
+import sys
+
+from distutils.version import LooseVersion
+
+from refresh_source import Refresher
+from spinnaker.run import run_quick
+
+
+class CommitTag:
+  """Provides a model class to capture the output of 'git show-ref --tags'.
+
+  We also capture the tag versions using `distutils.version` for easy semantic
+  version comparison for sorting.
+  """
+  def __init__(self, ref_line):
+    # ref_line is in the form "$commit_hash refs/tags/$tag"
+    tokens = ref_line.split(' ')
+    self.__hash = tokens[0]
+    tag_parts = tokens[1].split('/')
+    self.__tag = tag_parts[len(tag_parts) - 1]
+    self.__version = LooseVersion(self.__tag)
+
+  def __repr__(self):
+    return 'hash: %s, tag: %s, version: %s' % (self.__hash, self.__tag, self.__version)
+
+  @property
+  def hash(self):
+    return self.__hash
+
+  @property
+  def tag(self):
+    return self.__tag
+
+  @property
+  def version(self):
+    return self.__version
+
+
+class CommitMessage:
+  """Provides a model class to capture the output of 'git log --pretty'.
+  """
+  def __init__(self, hash, msg):
+    self.__hash = hash
+    self.__msg = msg
+
+  def __repr__(self):
+    return 'hash: %s, message: %s' % (self.__hash, self.__msg)
+
+  @property
+  def hash(self):
+    return self.__hash
+
+  @property
+  def msg(self):
+    return self.__msg
+
+
+class GitTagMissingException(Exception):
+  """Exception for misconfigured git tags in the operating repository."""
+  def __init__(self, message):
+    self.message = message
+
+
+class Annotator(Refresher):
+  """Provides semantic version tagging for Spinnaker repositories.
+
+  Each Spinnaker repository has tags that denote releases. These tags follow
+  semantic versioning. At the present time, there are two sets of tags in use
+  for the Spinnaker repositories: 'vX.Y.Z' for Netflix releases and 'sX.Y.Z'
```

Comment:
'sX.Y.Z' is what I was suggesting as a Spinnaker product version tag. We could start with our own version number, and not have any interaction with the existing `vX.Y.Z` tags at all.

The issue with using a 'v*' tag is how gradle resolves the tag to use for the next version. When we use `-PuseLastTag=true` in gradle, it will sort the tags of the form 'X.Y.Z' and 'vX.Y.Z*' (including '-rc.*') and use the latest tag by semver sort, favoring tags without the '-rc.*' if there's a collision.

If Netflix uses the '-rc.*' namespace, and we use the non-rc namespace and keep rolling the current microservice versions forward in the Spinnaker product releases I believe these issues wind up being resolved. We can use 'vX.Y.Z-$build' to publish unique artifacts to our private repo when testing, and push a 'vX.Y.Z' tag to master on a formal validated release, which TravisCI will ignore.

@ewiseblatt @cfieber @ajordens What do you think? Any issues/misconceptions here? Is Netflix willing to use the '-rc.*' namespace for internal releases?